### PR TITLE
Added Product to MySQL

### PR DESCRIPTION
**Pull request by Rudiachenko**

### CODE REVIEW

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            PreparedStatement statement = connection.prepareStatement(query,
+                    PreparedStatement.RETURN_GENERATED_KEYS);
+            statement.setString(1, item.getName());
+            statement.setBigDecimal(2, item.getPrice());
+            statement.executeUpdate();
+            ResultSet generatedKeys = statement.getGeneratedKeys();
+            while (generatedKeys.next()) {
+                long productId = generatedKeys.getLong(1);
+                item.setId(productId);
+            }
```
Comment (andromash) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java:32]: you can make a return already here

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM products "
+                    + "WHERE product_id = ?;";
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            resultSet.next();
+            Product product = createProduct(resultSet);
+            return Optional.of(product);
+
```
Comment (andromash) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    public List<Product> getAll() {
+        List<Product> productsList = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM products;";
+            PreparedStatement statement = connection.prepareStatement(query);
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                Product product = createProduct(resultSet);
+                productsList.add(product);
+            }
```
Comment (andromash) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java:68]: you can make return right here

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            statement.setBigDecimal(1, BigDecimal.valueOf(id));
+            statement.execute();
+            return true;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Delete of product with id "
+                    + id + " is failed", e);
+        }
+    }
+
+    private Product createProduct(ResultSet resultSet) throws SQLException {
```
Comment (andromash) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: does this name describe fully what it does?

File: src/main/resources/init_db.sql [hunk 5]
```diff
@@ -0,0 +1,10 @@
+CREATE SCHEMA `internet_shop` DEFAULT CHARACTER SET utf8 ;
+
+CREATE TABLE `internet_shop`.`products` (
+  `product_id` BIGINT(11) NOT NULL AUTO_INCREMENT,
+  `name` VARCHAR(45) NOT NULL,
```
Comment (andromash) [inline src/main/resources/init_db.sql]: is this size enough?

**Review by andromash — CHANGES_REQUESTED** [review]
Good job!

File: src/main/java/com/internet/shop/application/Main.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package com.internet.shop.application;
+
+import com.internet.shop.dao.jdbc.ProductDaoJdbcImpl;
+import com.internet.shop.model.Product;
+
+public class Main {
+    public static void main(String[] args) {
+        ProductDaoJdbcImpl productDaoJdbc = new ProductDaoJdbcImpl();
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/application/Main.java]: - let's use injector)
- we can remove dao annotation above ProductDaoImpl class

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Dao
+public class ProductDaoJdbcImpl implements ProductDao {
+    @Override
+    public Product create(Product item) {
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
    public Product create(Product product) {
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "INSERT INTO products (name, price) "
+                    + "VALUES (?, ?);";
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            String query = "INSERT INTO products (name, price) VALUES (?, ?);";
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            throw new DataProcessingException("It is impossible to create a product: "
+                    + item, e);
+        }
+    }
+
+    @Override
+    public Optional<Product> getById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM products "
+                    + "WHERE product_id = ?;";
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            String query = "SELECT * FROM products WHERE product_id = ?;";
```
should we get deleted a product?

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            resultSet.next();
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: what does happen if resultSet.next() == null?

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+            throw new DataProcessingException("Get product with id "
+                    + id + " is failed", e);
+        }
+    }
+
+    @Override
+    public List<Product> getAll() {
+        List<Product> productsList = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM products;";
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: how about the deleted products?

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+            return productsList;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Get products is failed", e);
+        }
+    }
+
+    @Override
+    public Product update(Product item) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "UPDATE products "
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: - can we update deleted data?
- let's combine 3 lines into two or one

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    @Override
+    public boolean delete(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "UPDATE products "
+                    + "SET deleted = 1 "
+                    + "WHERE product_id = ?;";
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setBigDecimal(1, BigDecimal.valueOf(id));
+            statement.execute();
+            return true;
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: should we always return true? let's check the result of statement.executeUpdate() and return the result of checking

File: src/main/resources/init_db.sql [hunk 14]
```diff
@@ -0,0 +1,10 @@
+CREATE SCHEMA `internet_shop` DEFAULT CHARACTER SET utf8 ;
+
+CREATE TABLE `internet_shop`.`products` (
+  `product_id` BIGINT(11) NOT NULL AUTO_INCREMENT,
+  `name` VARCHAR(225) NOT NULL,
+  `price` BIGINT(11) NOT NULL,
+  `deleted` TINYINT(11) NULL DEFAULT 0,
+  PRIMARY KEY (`product_id`),
+  UNIQUE INDEX `product_id_UNIQUE` (`product_id` ASC) VISIBLE,
+  UNIQUE INDEX `name_UNIQUE` (`name` ASC) VISIBLE);
```
Comment (Sofasmile) [inline src/main/resources/init_db.sql]: should product_name has UNIQUE INDEX and product_id too?

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**